src/microbe_stage/ProcessSystem.cs

```diff
@@ -244,18 +244,16 @@ private static float GetDissolvedInBiome(Compound compound, BiomeConditions biom
 
             // Environmental compound that can limit the rate
 
-            var input = new ProcessSpeedInformation.EnvironmentalInput(entry.Key, entry.Value);
-
             var availableInEnvironment = GetDissolvedInBiome(entry.Key, biome);
 
-            input.AvailableAmount = availableInEnvironment;
+            result.AvailableAmounts[entry.Key] = availableInEnvironment;
 
             // More than needed environment value boosts the effectiveness
-            input.AvailableRate = availableInEnvironment / entry.Value;
+            result.AvailableRates[entry.Key] = availableInEnvironment / entry.Value;
 
-            speedFactor *= input.AvailableRate;
+            speedFactor *= result.AvailableRates[entry.Key];
```

Comment:
unnecessary read back from the dictionary `availableInEnvironment / entry.Value` should be stored in a variable so it can be used in multiple places here.